### ISSUE TITLE
PageTitle - Wrong instructions re migrating Blazor WebAssembly to BWA

### ISSUE THREAD

**Issue by hakenr** [dotnet/AspNetCore.Docs#32521]
### Description

In the section mentioned here: https://learn.microsoft.com/en-us/aspnet/core/migration/70-80?view=aspnetcore-8.0&tabs=visual-studio#convert-a-hosted-blazor-webassembly-app-into-a-blazor-web-app

The suggested instruction to replace
```html
<title>...</title>
```

with
```razor
<HeadOutlet @rendermode="new InteractiveWebAssemblyRenderMode(prerender: false)" />
<PageTitle>...</PageTitle>
```
may not be entirely accurate. The `PageTitle` component does not have any effect in this context and won't render anything, neither in Static SSR nor in WASM.

The original purpose of the `<title>...</title>` element in `index.html` was to provide a default title for all pages that don't have a local `<PageTitle>...</PageTitle>` override (though such an override can be placed in `MainLayout.razor`, its effect only becomes visible after the WASM loads, not during the app's initial loading phase).

To maintain the original functionality of `<title>...</title>` from `index.html`, the correct content might instead be:
```razor
<HeadOutlet @rendermode="new InteractiveWebAssemblyRenderMode(prerender: false)" />
<title>...</title>
```

If this adjustment seems appropriate, I'm ready to prepare a PR to correct the instructions.
There are also alternative methods to achieve the same effect.

cc @guardrex

PS: To reproduce
1. Create a new BWA in Visual Studio (WASM global interactivity).
2. Go to `App.razor` and apply this change:
```diff
-<HeadOutlet @rendermode="InteractiveWebAssembly" />
+<HeadOutlet @rendermode="new InteractiveWebAssemblyRenderMode(prerender: false)" />
+<PageTitle>MyTitle</PageTitle>
```
3. Go to `Home.index` and remove the `<PageTitle>Home</PageTitle>`
4. Start the app, the will be no title set for the Home page.

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/migration/70-80?view=aspnetcore-8.0&tabs=visual-studio#convert-a-hosted-blazor-webassembly-app-into-a-blazor-web-app

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/migration/70-80.md

### Document ID

534ab561-09c5-c33b-4394-0effba34d666

### Article author

@Rick-Anderson

**Comment by guardrex** [comment]
Morning @hakenr ☕ ... The BWA project template doesn't have a `<title>` element in the `App` component ...

https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/App.razor#L5-L26

... and we were trying to get devs converted to an exact BWA OOB app. Therefore, guidance for retaining `<title>` from `index.html` wasn't included. Also ... IIRC 🤔 ... they had a reason for not retaining it, and I seem to recall that when I tested with it there was some funky artifact ... a failure or strange 👽 undesirable behavior ... with using a `<title>` tag, but I don't recall what happened.

They have at least one open issue for page titles 👉 https://github.com/dotnet/aspnetcore/issues/50268 👈 that I don't think deals with what you've reporting here. They might have one or more additional title-related issues.

Since we're trying to match what the current template produces, your concern should to go to the product unit for investigation/resolution. They may decide that it's ok to break from the template in the guidance, but they'll need to make that call. Otherwise, your PU issue might lead to a change in the project template, perhaps even for .NET 9.

Open for them on their repo ...

https://github.com/dotnet/aspnetcore/issues

... to ask about it, and please add the ping/cross-link back here in case they determine that they'd like to address something in the doc on it ...

```
cc: @guardrex https://github.com/dotnet/AspNetCore.Docs/issues/32521
```

**Comment by hakenr** [comment]
@guardrex The [guidance](https://learn.microsoft.com/en-us/aspnet/core/migration/70-80?view=aspnetcore-8.0&tabs=visual-studio#convert-a-hosted-blazor-webassembly-app-into-a-blazor-web-app) I'm referencing suggests adding `<PageTitle>...</PageTitle>` to the new `App.razor` component. This line is not part of the BWA template and does not work.

Please let's not focus on my workaround (adding `<title>...</title>` to `App.razor`). Instead, let's address the problem in the guidance where it suggests adding something that is neither aligned with the BWA template nor functional.

If my code change doesn’t appeal to you (no worries, it might have issues), and if there's no better solution, perhaps we could update the guidance to clearly state that "setting the default page title, which is used during the page startup phase (initial HTML served from the server), is not supported in BWA Interactive WebAssembly without prerendering" (if that is the case?). It would be more helpful than recommending `<PageTitle>...</PageTitle>`, which does not work.

**Comment by guardrex** [comment]
Sorry ... yes ... I did just notice that. I was too focused on the `<title>` workaround piece, which I think presented a problem.

Let's cross-link to the layout approach because that's what the PU favors. Here's the link ...

```
<xref:blazor/components/control-head-content?view=aspnetcore-8.0#set-a-page-title-for-components-via-a-layout
```

**Comment by hakenr** [comment]
In this scenario, the layout functions as an interactive client component, rendering the title only after the WASM runtime starts. This differs from the original behavior, where the title was set in the initial HTML served by the server. Should we mention this change in the guidance?
(There are several other features dropped from the original Blazor WASM setup, like the loading progress indicator, that aren't mentioned by the guidance either. Just wondering if we should highlight this `title`-aspect of the migration.)

**Comment by guardrex** [comment]
Yes! I think these are great recommendations so that devs aren't surprised. I'm sorry that there wasn't more time to explain these aspects when the content was written. Even rushing, I wasn't finished at GA with .NET 8 updates (and I'm still not finished). That's the magnitude of the shift in docs for the .NET 8 release.

**Comment by guardrex** [comment]
... and BTW ... it wasn't done in a total vacuum ... Mackinnon signed off on this guidance here ...

https://github.com/dotnet/AspNetCore.Docs/pull/32144

... and very few 🦖 ***RexHacks***&trade; 🦖 get past his watchful 👁️! 😆

**Comment by hakenr** [comment]
>>Create a new BWA in Visual Studio (WASM global interactivity).
>
> ... but that's not the migration path for this section.

This wasn't intended as a migration scenario, but rather as proof that the `<PageTitle>...</PageTitle>` component in `App.razor` of a BWA WebAssembly application without prerendering does not function correctly.

> I do see another change to make. I'm going to go ahead and fix this myself right now, but ping me if you've already started and I'll stop. I think our timezone differences have it late there, so you might be 🛌💤 now.

No, I haven't started working on this yet. Since my original solution wasn't approved, I'll leave this up to you unless you specifically request that I modify the guidance in a certain way.
BTW, I'm in the UTC+2 timezone (CEST, Prague), so it's 4:20pm here in Europe right now. What is your timezone?

**Comment by guardrex** [comment]
I'll set my PR up as an "addresses" PR, so you to make further modifications that you think would be helpful. For example, the loss of the loading progress indicator could link over to ...

https://learn.microsoft.com/aspnet/core/blazor/fundamentals/startup#client-side-loading-progress-indicators

```
<xref:blazor/fundamentals/startup#client-side-loading-progress-indicators>
```

... with some explanation.

Plus, whatever else you think would be helpful and any changes to what I'm doing now.

> UTC+2 timezone (CEST, Prague)

I see ... not late then. I think it's UTC−04:00 [US east coast (EST)]. I'm in Melbourne, Florida, so I'm on the "Space Coast" with the bonus of many 🚀 to watch blasting off. I wish I could get to Europe. Prague is ***definitely*** on my list of cities to visit. It looks like such travel will need to wait until retirement.

**Comment by guardrex** [comment]
@hakenr ... Ok ... that's done. I recommend taking it from there with further modifications, including to what I just did. I'm (as usual) ***rushing a bit*** 🏃‍♂️ with that to get on with other work. Thanks for this issue! Great 👁️ as usual on catching problems like this! 🏅

**Comment by hakenr** [comment]
@guardrex Probably not - it's "just" a guidance on how to migrate from one project setup to another, and this might not be the best place to detail all the differences between them. A cross-link could be appropriate, but there probably isn't an article that explains all the differences and limitations... :-D
